defi/src/storeTvlInterval/computeTVL.ts

```diff
@@ -1,98 +1,110 @@
 import BigNumber from "bignumber.js";
 import fetch from "node-fetch";
-import { addStaleCoin, StaleCoins } from "./staleCoins";
-
 const ethereumAddress = "0x0000000000000000000000000000000000000000";
 const weth = "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2";
-const DAY = 24 * 3600;
+
 type Balances = {
   [symbol: string]: number;
 };
 
-export default async function (balances: { [address: string]: string }, timestamp: "now" | number, staleCoins:StaleCoins) {
+function normalizeBalances(balances: { [address: string]: string }) {
+  Object.keys(balances).map((key) => {
+    if (+balances[key] === 0) {
+      delete balances[key];
+      return;
+    }
+
+    const normalisedKey = key.startsWith("0x")
+      ? `ethereum:${key.toLowerCase()}`
+      : !key.includes(":")
+      ? `coingecko:${key.toLowerCase()}`
+      : key;
+
+    // sols case sensitive so no normalising
+    if (key == normalisedKey || key.startsWith("solana:")) return;
+
+    balances[normalisedKey] =
+      normalisedKey in balances
+        ? (Number(balances[normalisedKey]) + Number(balances[key])).toString()
+        : balances[key];
+
+    delete balances[key];
+  });
+
   const eth = balances[ethereumAddress];
   if (eth !== undefined) {
     balances[weth] = new BigNumber(balances[weth] ?? 0).plus(eth).toFixed(0);
     delete balances[ethereumAddress];
   }
-  const PKsToTokens = {} as { [t: string]: string[] };
-  const readKeys = Object.keys(balances)
-    .map((address) => {
-      let prefix = "";
-      if(address.startsWith("0x")){
-        prefix = "ethereum:"
-      } else if(!address.includes(":")){
-        prefix = "coingecko:"
-      }
-      let normalizedAddress = address.toLowerCase()
-      if(address.startsWith("solana:")){
-        normalizedAddress = address
-      }
-      const PK = `${prefix}${normalizedAddress}`;
-      if (PKsToTokens[PK] === undefined) {
-        PKsToTokens[PK] = [address];
-        return PK;
-      } else {
-        PKsToTokens[PK].push(address);
-        return undefined;
-      }
-    })
-    .filter((item) => item !== undefined) as string[];
+
+  return balances;
+}
+
+async function fetchTokenData(
+  timestamp: "now" | number,
+  balances: { [address: string]: string }
+) {
   const readRequests = [];
-  for (let i = 0; i < readKeys.length; i += 100) {
-    const body = {
-      "coins": readKeys.slice(i, i + 100),
-    } as any
-    if(timestamp !== "now"){
-      body.timestamp = timestamp;
-    }
+  const burl = "https://coins.llama.fi/prices/";
+  const historical =
+    timestamp == "now" ? "current/" : `historical/${timestamp}/`;
+
+  for (let i = 0; i < Object.keys(balances).length; i += step) {
+    const coins = Object.keys(balances)
+      .slice(i, i + step)
+      .reduce((p, c) => p + `${c},`, "");
```

Comment:
.join(',') is much more efficient and readable than that reduce

appending string by string in a string that keeps growing is extremely slow, since it ends up with multiple memory allocations

something like this:
1. allocate memory for string1
2. do string = string1+string2, now you need more space so you allocate space for the new string, copy string1 there and then copy string2
3. do string += string3, same again: new memory allocated and whole old string is copied there

this is generally super slow since you need to be reallocating ever growing strings

idk what join() does but i guess it calculates the final length of the string, allocates it beforehand and then copies all the strings there